File: app/src/main/java/com/kidozh/discuzhub/activities/AddIntroActivity.java
package com.kidozh.discuzhub.activities;

import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Paint;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.Toast;

import com.kidozh.discuzhub.R;
import com.kidozh.discuzhub.adapter.UrlSuggestionAdapter;
import com.kidozh.discuzhub.database.BBSInformationDatabase;
import com.kidozh.discuzhub.databinding.ActivityBbsAddIntroBinding;
import com.kidozh.discuzhub.entities.SuggestURLInfo;
import com.kidozh.discuzhub.entities.bbsInformation;
import com.kidozh.discuzhub.results.AddCheckResult;
import com.kidozh.discuzhub.utilities.AnimationUtils;
import com.kidozh.discuzhub.utilities.VibrateUtils;
import com.kidozh.discuzhub.utilities.bbsParseUtils;
import com.kidozh.discuzhub.utilities.URLUtils;
import com.kidozh.discuzhub.utilities.NetworkUtils;
import com.kidozh.discuzhub.viewModels.AddBBSViewModel;

import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

import es.dmoral.toasty.Toasty;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class AddIntroActivity extends BaseStatusActivity
        implements UrlSuggestionAdapter.OnClickSuggestionListener {
    private static final String TAG = AddIntroActivity.class.getSimpleName();


    UrlSuggestionAdapter adapter;

    AddBBSViewModel viewModel;
    ActivityBbsAddIntroBinding binding;



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityBbsAddIntroBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        
        viewModel = new ViewModelProvider(this).get(AddBBSViewModel.class);
        configureRecyclerview();
        bindViewModel();
        configureUrlEditText();
        configureContinueBtn();
        configureAddGuide();
    }

    private void configureAddGuide(){
        binding.bbsAddGuide.setPaintFlags(binding.bbsAddGuide.getPaintFlags() | Paint.UNDERLINE_TEXT_FLAG);
        binding.bbsAddGuide.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                Uri uri = Uri.parse("https://discuzhub.kidozh.com/add-a-bbs-guide/");
                Intent intent = new Intent(Intent.ACTION_VIEW, uri);
                startActivity(intent);
            }
        });
    }

    private void configureRecyclerview(){
        adapter = new UrlSuggestionAdapter();
        binding.bbsAddIntroRecyclerview.setLayoutManager(new LinearLayoutManager(this));
        binding.bbsAddIntroRecyclerview.setHasFixedSize(true);
        binding.bbsAddIntroRecyclerview.setAdapter(AnimationUtils.INSTANCE.getAnimatedAdapter(this,adapter));
        binding.bbsAddIntroRecyclerview.setItemAnimator(AnimationUtils.INSTANCE.getRecyclerviewAnimation(this));
        // add examples
        List<SuggestURLInfo> suggestURLInfoList = new ArrayList<>();
        suggestURLInfoList.add(new SuggestURLInfo("https://bbs.nwpu.edu.cn",getString(R.string.bbs_url_example_npubbs),true));
        suggestURLInfoList.add(new SuggestURLInfo("https://bbs.comsenz-service.com",getString(R.string.bbs_url_example_discuz_support),true));
        suggestURLInfoList.add(new SuggestURLInfo("https://www.mcbbs.net",getString(R.string.bbs_url_example_mcbbs),true));
        suggestURLInfoList.add(new SuggestURLInfo("https://keylol.com",getString(R.string.bbs_url_example_keylol),true));
        suggestURLInfoList.add(new SuggestURLInfo("https://bbs.qzzn.com",getString(R.string.bbs_url_example_qzzn),true));
        suggestURLInfoList.add(new SuggestURLInfo("https://www.right.com.cn/forum",getString(R.string.bbs_url_example_right_com),true));
        adapter.setSuggestURLInfoList(suggestURLInfoList);
    }

    private void bindViewModel(){
        viewModel.currentURLLiveData.observe(this, new Observer<String>() {
            @Override
            public void onChanged(String s) {
                // need to analyze the URL
                List<SuggestURLInfo> suggestURLInfos = getSuggestedURLList(s);
                adapter.setSuggestURLInfoList(suggestURLInfos);
            }
        });
        viewModel.useSafeClientLiveData.observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean aBoolean) {
                adapter.setUseSafeClient(aBoolean);
            }
        });
        viewModel.isLoadingLiveData.observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean aBoolean) {
                if(aBoolean){
                    binding.bbsAddIntroProgressBar.setVisibility(View.VISIBLE);
                }
                else {
                    binding.bbsAddIntroProgressBar.setVisibility(View.GONE);
                }
            }
        });
        viewModel.errorTextLiveData.observe(this, new Observer<String>() {
            @Override
            public void onChanged(String s) {
                if(s.length()!=0){
                    Toasty.warning(getApplication(),s, Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    private List<SuggestURLInfo> getSuggestedURLList(String urlString){
        List<SuggestURLInfo> suggestURLInfoList = new ArrayList<>();
        if(urlString.equals("")){
            suggestURLInfoList.add(new SuggestURLInfo("https://bbs.nwpu.edu.cn",getString(R.string.bbs_url_example_npubbs),true));
            suggestURLInfoList.add(new SuggestURLInfo("https://bbs.comsenz-service.com",getString(R.string.bbs_url_example_discuz_support),true));
            suggestURLInfoList.add(new SuggestURLInfo("https://www.mcbbs.net",getString(R.string.bbs_url_example_mcbbs),true));
            suggestURLInfoList.add(new SuggestURLInfo("https://keylol.com",getString(R.string.bbs_url_example_keylol),true));
            suggestURLInfoList.add(new SuggestURLInfo("https://www.1point3acres.com/bbs",getString(R.string.bbs_url_example_1point3acres),true));
            suggestURLInfoList.add(new SuggestURLInfo("https://www.right.com.cn/forum",getString(R.string.bbs_url_example_right_com),true));
        }
        else {
            // add url
            try{
                URL url = new URL(urlString);
                String[] splitString = urlString.split("/");
                if(splitString.length >=3){
                    String pathPrefix = splitString[0]+"/"+splitString[1]+"/"+splitString[2];

                    Log.d(TAG,"Path prefix "+pathPrefix);
                    suggestURLInfoList.add(new SuggestURLInfo(pathPrefix,getString(R.string.bbs_url_suggestion_host),false));
                    for(int i=3;i<splitString.length;i++){
                        pathPrefix += "/" + splitString[i];
                        suggestURLInfoList.add(new SuggestURLInfo(pathPrefix,getString(R.string.bbs_url_suggestion_level, i-2),false));
                    }
                }
            }
            catch (Exception e){
                suggestURLInfoList.add(new SuggestURLInfo("https://bbs.nwpu.edu.cn",getString(R.string.bbs_url_example_npubbs),true));
                suggestURLInfoList.add(new SuggestURLInfo("https://bbs.comsenz-service.com",getString(R.string.bbs_url_example_discuz_support),true));
                suggestURLInfoList.add(new SuggestURLInfo("https://www.mcbbs.net",getString(R.string.bbs_url_example_mcbbs),true));
                suggestURLInfoList.add(new SuggestURLInfo("https://keylol.com",getString(R.string.bbs_url_example_keylol),true));
                suggestURLInfoList.add(new SuggestURLInfo("https://www.1point3acres.com/bbs",getString(R.string.bbs_url_example_1point3acres),true));
                suggestURLInfoList.add(new SuggestURLInfo("https://www.right.com.cn/forum",getString(R.string.bbs_url_example_right_com),true));
            }


        }
        return suggestURLInfoList;
    }

    private void configureUrlEditText(){
        binding.bbsAddIntroUrlEdittext.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // need to post data
                viewModel.currentURLLiveData.setValue(s.toString());

            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        binding.bbsAddIntroHttpsCheckbox.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                viewModel.useSafeClientLiveData.setValue(isChecked);
            }
        });
    }

    private void configureContinueBtn(){
        binding.bbsAddIntroContinueButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String urlString = binding.bbsAddIntroUrlEdittext.getText().toString();
                try{
                    URL url = new URL(urlString);
                    queryBBSInfo(urlString);
                }
                catch (MalformedURLException e){
                    Toasty.warning(getApplication(),getString(R.string.add_bbs_url_failed, urlString),Toast.LENGTH_LONG).show();
                }

            }
        });
    }


    private void queryBBSInfo(String base_url){
        boolean useSafeClient = binding.bbsAddIntroHttpsCheckbox.isChecked();
        viewModel.isLoadingLiveData.postValue(true);

        URLUtils.setBaseUrl(base_url);
        String query_url = URLUtils.getBBSForumInformationUrl();
        OkHttpClient client = NetworkUtils.getPreferredClient(this,useSafeClient);
        Request request;
        try{
            URL url = new URL(query_url);
//            if(url.getAuthority()!=null && url.getHost()!=null){
//                throw new Exception();
//            }
            request = new Request.Builder().url(query_url).build();
        }
        catch (Exception e){
            viewModel.isLoadingLiveData.postValue(false);
            viewModel.errorTextLiveData.postValue(getString(R.string.bbs_base_url_invalid));
            e.printStackTrace();
            return;
        }

        Call call = client.newCall(request);
        Log.d(TAG,"Query check URL "+query_url);
        Activity activity = this;
        call.enqueue(new Callback() {

            @Override
            public void onFailure(Call call, IOException e) {
                viewModel.isLoadingLiveData.postValue(false);
                e.printStackTrace();
                viewModel.errorTextLiveData.postValue(getString(R.string.network_failed));

            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                viewModel.isLoadingLiveData.postValue(false);
                String s;
                if(response.isSuccessful() && response.body()!=null){
                    s = response.body().string();
                    Log.d(TAG,"check response " +s);
                    AddCheckResult checkResult = bbsParseUtils.parseCheckInfoResult(s);
                    if(checkResult!=null){
                        bbsInformation bbsInfo = checkResult.toBBSInformation(base_url);
                        new addNewForumInformationTask(bbsInfo,activity).execute();
                        finishAfterTransition();
                    }
                    else {
                        viewModel.errorTextLiveData.postValue(getString(R.string.parse_failed));
                    }
                }
                else {
                    viewModel.errorTextLiveData.postValue(getString(R.string.parse_failed));
                }
            }
        });
    }

    public static class addNewForumInformationTask extends AsyncTask<Void, Void, Void> {
        private bbsInformation forumInfo;
        private Activity activity;
        public addNewForumInformationTask(bbsInformation bbsInformation, Activity activity){
            this.forumInfo = bbsInformation;
            this.activity = activity;
            forumInfo.position = 9999999;
        }
        @Override
        protected Void doInBackground(Void... voids) {
            BBSInformationDatabase
                    .getInstance(activity)
                    .getForumInformationDao().insert(forumInfo);


            Log.d(TAG, "add forum into database"+forumInfo.site_name);
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);
            Toasty.success(activity,
                    activity.getString(R.string.add_a_bbs_successfully,forumInfo.site_name),
                    Toast.LENGTH_SHORT).show();
            activity.finishAfterTransition();
        }
    }

    @Override
    public void onClickSuggestion(SuggestURLInfo suggestURLInfo) {
        VibrateUtils.vibrateForClick(this);
        // lazy load
        if(!binding.bbsAddIntroUrlEdittext.getText().toString().equals(suggestURLInfo.url)){
            binding.bbsAddIntroUrlEdittext.setText(suggestURLInfo.url);
        }

    }
}

File: app/src/main/java/com/kidozh/discuzhub/activities/AddIntroActivity.kt
package com.kidozh.discuzhub.activities

import com.kidozh.discuzhub.utilities.AnimationUtils.getAnimatedAdapter
import com.kidozh.discuzhub.utilities.AnimationUtils.getRecyclerviewAnimation

import com.kidozh.discuzhub.activities.BaseStatusActivity
import com.kidozh.discuzhub.adapter.UrlSuggestionAdapter.OnClickSuggestionListener
import com.kidozh.discuzhub.adapter.UrlSuggestionAdapter
import com.kidozh.discuzhub.viewModels.AddBBSViewModel
import android.os.Bundle
import androidx.lifecycle.ViewModelProvider
import android.content.Intent
import android.graphics.Paint
import android.net.Uri
import androidx.recyclerview.widget.LinearLayoutManager
import com.kidozh.discuzhub.entities.SuggestURLInfo
import com.kidozh.discuzhub.R
import es.dmoral.toasty.Toasty
import android.widget.Toast
import com.kidozh.discuzhub.entities.Discuz
import com.kidozh.discuzhub.database.DiscuzDatabase
import com.kidozh.discuzhub.activities.AddIntroActivity
import android.text.TextWatcher
import android.text.Editable
import android.util.Log
import android.view.View
import android.widget.CompoundButton
import androidx.lifecycle.Observer
import com.kidozh.discuzhub.databinding.ActivityBbsAddIntroBinding
import com.kidozh.discuzhub.utilities.VibrateUtils
import java.lang.Exception
import java.net.MalformedURLException
import java.net.URL
import java.util.ArrayList

class AddIntroActivity : BaseStatusActivity(), OnClickSuggestionListener {
    var adapter: UrlSuggestionAdapter = UrlSuggestionAdapter()
    lateinit var viewModel: AddBBSViewModel
    var binding: ActivityBbsAddIntroBinding? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityBbsAddIntroBinding.inflate(layoutInflater)
        setContentView(binding!!.root)
        viewModel = ViewModelProvider(this).get(AddBBSViewModel::class.java)
        configureRecyclerview()
        bindViewModel()
        configureUrlEditText()
        configureContinueBtn()
        configureAddGuide()
    }

    private fun configureAddGuide() {
        binding!!.bbsAddGuide.paintFlags = binding!!.bbsAddGuide.paintFlags or Paint.UNDERLINE_TEXT_FLAG
        binding!!.bbsAddGuide.setOnClickListener {
            val uri = Uri.parse("https://discuzhub.kidozh.com/add-a-bbs-guide/")
            val intent = Intent(Intent.ACTION_VIEW, uri)
            startActivity(intent)
        }
    }

    private fun configureRecyclerview() {

        binding!!.bbsAddIntroRecyclerview.layoutManager = LinearLayoutManager(this)
        binding!!.bbsAddIntroRecyclerview.setHasFixedSize(true)
        binding!!.bbsAddIntroRecyclerview.adapter = getAnimatedAdapter(this, adapter)
        binding!!.bbsAddIntroRecyclerview.itemAnimator = getRecyclerviewAnimation(this)
        // add examples
        val suggestURLInfoList: MutableList<SuggestURLInfo> = ArrayList()
        suggestURLInfoList.add(SuggestURLInfo("https://bbs.nwpu.edu.cn", getString(R.string.bbs_url_example_npubbs), true))
        suggestURLInfoList.add(SuggestURLInfo("https://bbs.comsenz-service.com", getString(R.string.bbs_url_example_discuz_support), true))
        suggestURLInfoList.add(SuggestURLInfo("https://www.mcbbs.net", getString(R.string.bbs_url_example_mcbbs), true))
        suggestURLInfoList.add(SuggestURLInfo("https://keylol.com", getString(R.string.bbs_url_example_keylol), true))
        suggestURLInfoList.add(SuggestURLInfo("https://bbs.qzzn.com", getString(R.string.bbs_url_example_qzzn), true))
        suggestURLInfoList.add(SuggestURLInfo("https://www.right.com.cn/forum", getString(R.string.bbs_url_example_right_com), true))
        adapter.setSuggestURLInfoList(suggestURLInfoList)
    }

    private fun bindViewModel() {
        viewModel.currentURLLiveData.observe(this, { s -> // need to analyze the URL
            val suggestURLInfos = getSuggestedURLList(s)
            adapter.setSuggestURLInfoList(suggestURLInfos)
        })
        viewModel.isLoadingLiveData.observe(this, { aBoolean ->
            if (aBoolean) {
                binding!!.bbsAddIntroProgressBar.visibility = View.VISIBLE
            } else {
                binding!!.bbsAddIntroProgressBar.visibility = View.GONE
            }
        })
        viewModel.errorTextLiveData.observe(this, { s ->
            if (s.length != 0) {
                Toasty.warning(application, s, Toast.LENGTH_SHORT).show()
            }
        })
        viewModel.verifiedBBS.observe(this, Observer { bbsInformation: Discuz? ->
            if (bbsInformation != null) {
                Thread {
                    DiscuzDatabase
                            .getInstance(this)
                            .forumInformationDao.insert(bbsInformation)
                }.start()
                Toasty.success(this,
                        getString(R.string.add_a_bbs_successfully, bbsInformation.site_name),
                        Toast.LENGTH_SHORT).show()
                finishAfterTransition()
            }
        })
        viewModel.errorMessageMutableLiveData.observe(this, { it ->
            if(it != null){
                Toasty.error(this,getString(R.string.discuz_api_message_template,it.key,it.content),Toast.LENGTH_SHORT).show()
            }

        })
    }

    private fun getSuggestedURLList(urlString: String): List<SuggestURLInfo> {
        val suggestURLInfoList: MutableList<SuggestURLInfo> = ArrayList()
        if (urlString == "") {
            suggestURLInfoList.add(SuggestURLInfo("https://bbs.nwpu.edu.cn", getString(R.string.bbs_url_example_npubbs), true))
            suggestURLInfoList.add(SuggestURLInfo("https://bbs.comsenz-service.com", getString(R.string.bbs_url_example_discuz_support), true))
            suggestURLInfoList.add(SuggestURLInfo("https://www.mcbbs.net", getString(R.string.bbs_url_example_mcbbs), true))
            suggestURLInfoList.add(SuggestURLInfo("https://keylol.com", getString(R.string.bbs_url_example_keylol), true))
            suggestURLInfoList.add(SuggestURLInfo("https://bbs.qzzn.com", getString(R.string.bbs_url_example_qzzn), true))
            suggestURLInfoList.add(SuggestURLInfo("https://www.right.com.cn/forum", getString(R.string.bbs_url_example_right_com), true))
        } else {
            // add url
            try {
                val url = URL(urlString)
                val splitString = urlString.split("/").toTypedArray()
                if (splitString.size >= 3) {
                    var pathPrefix = splitString[0] + "/" + splitString[1] + "/" + splitString[2]
                    Log.d(TAG, "Path prefix $pathPrefix")
                    suggestURLInfoList.add(SuggestURLInfo(pathPrefix, getString(R.string.bbs_url_suggestion_host), false))
                    for (i in 3 until splitString.size) {
                        pathPrefix += "/" + splitString[i]
                        suggestURLInfoList.add(SuggestURLInfo(pathPrefix, getString(R.string.bbs_url_suggestion_level, i - 2), false))
                    }
                }
            } catch (e: Exception) {
                suggestURLInfoList.add(SuggestURLInfo("https://bbs.nwpu.edu.cn", getString(R.string.bbs_url_example_npubbs), true))
                suggestURLInfoList.add(SuggestURLInfo("https://bbs.comsenz-service.com", getString(R.string.bbs_url_example_discuz_support), true))
                suggestURLInfoList.add(SuggestURLInfo("https://www.mcbbs.net", getString(R.string.bbs_url_example_mcbbs), true))
                suggestURLInfoList.add(SuggestURLInfo("https://keylol.com", getString(R.string.bbs_url_example_keylol), true))
                suggestURLInfoList.add(SuggestURLInfo("https://www.1point3acres.com/bbs", getString(R.string.bbs_url_example_1point3acres), true))
                suggestURLInfoList.add(SuggestURLInfo("https://www.right.com.cn/forum", getString(R.string.bbs_url_example_right_com), true))
            }
        }
        return suggestURLInfoList
    }

    private fun configureUrlEditText() {
        binding!!.bbsAddIntroUrlEdittext.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                // need to post data
                viewModel.currentURLLiveData.value = s.toString()
            }

            override fun afterTextChanged(s: Editable) {}
        })
        binding!!.autoAddBbs.setOnCheckedChangeListener { buttonView: CompoundButton?, isChecked: Boolean -> viewModel.autoVerifyURLLiveData.postValue(isChecked) }
    }

    private fun configureContinueBtn() {
        binding!!.bbsAddIntroContinueButton.setOnClickListener {
            val urlString = binding!!.bbsAddIntroUrlEdittext.text.toString()
            try {
                val url = URL(urlString)
                viewModel.verifyURL()
            } catch (e: MalformedURLException) {
                Toasty.warning(application, getString(R.string.add_bbs_url_failed, urlString), Toast.LENGTH_LONG).show()
            }
        }
    }

    override fun onClickSuggestion(suggestURLInfo: SuggestURLInfo) {
        VibrateUtils.vibrateForClick(this)
        // lazy load
        if (binding!!.bbsAddIntroUrlEdittext.text.toString() != suggestURLInfo.url) {
            binding!!.bbsAddIntroUrlEdittext.setText(suggestURLInfo.url)
        }
    }

    var hasSubmitAutoVerify = false
    override fun onURLVerified(base_url: String) {
        val autoCheck = viewModel.autoVerifyURLLiveData.value!!
        if (!hasSubmitAutoVerify && autoCheck) {
            hasSubmitAutoVerify = true
            // only implement when auto check is on
            // not triggering url
            viewModel.currentURLLiveData.postValue(base_url)
            viewModel.verifyURL()
        }
    }

    companion object {
        private val TAG = AddIntroActivity::class.java.simpleName
    }
}